pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -414,18 +414,28 @@ private void updateBundleData() {
 
             // Remove all loaded bundles from the preallocated maps.
             final Map<String, BundleData> preallocatedBundleData = brokerData.getPreallocatedBundleData();
-            // Should not iterate with more than one thread at a time.
             synchronized (preallocatedBundleData) {
-                final Iterator<Map.Entry<String, BundleData>> preallocatedIterator = preallocatedBundleData.entrySet()
-                        .iterator();
-                while (preallocatedIterator.hasNext()) {
-                    final String bundle = preallocatedIterator.next().getKey();
-                    if (bundleData.containsKey(bundle)) {
-                        preallocatedIterator.remove();
-                        preallocatedBundleToBroker.remove(bundle);
+                for (String bundleName : brokerData.getLocalData().getBundles()) {
+                    if (preallocatedBundleData.containsKey(bundleName)) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Preallocated bundle {[]} is now loaded on broker [{}]", bundleName, broker);
+                        }
+                        final Iterator<Map.Entry<String, BundleData>> preallocatedIterator = preallocatedBundleData.entrySet()
+                                .iterator();
+                        while (preallocatedIterator.hasNext()) {
```

Comment:
That's a good observation. I believe the runtime is closer to O(m _log n_) where m = number of loaded bundles on the broker and n is the number of preallocated bundles on the broker (since there are two nested loops of different sizes). In the normal case, m >> n and you are correct in observing it would be better to switch the outer loop to get O(n _log m_) runtime. However, when we are in this steady state, the function is not called that often (perhaps only once per 30 seconds).

On the other hand, when things are changing rapidly (like during a deployment), then the preallocated array is larger (but only for a minute or two) and n >> m so we do more work. 

I've tested it both ways and both work, it's just a matter of what to optimize for. I'm happy to change the outer loop (which was missing in the previous version altogether and nothing ever got removed).

One more subtle problem is this: If a broker loads a bundle and dies before it can get a load report out, I believe we will still be holding the preallocation and the bundle would get reassigned to a broker that is down. If that turns out to be true, we should TTL the broker-to-bundle preallocation map or remove all the broker's preallocations when the ephemeral broker node disappears.